fathom/src/core/semantics.rs

```diff
@@ -181,6 +181,13 @@ impl<'arena> Split<'arena> {
     }
 }
 
+#[derive(Clone, Debug)]
+pub enum SplitConstBranches<'arena> {
+    Const(Const, ArcValue<'arena>, Split<'arena>),
+    Default(Closure<'arena>),
+    None,
+}
```

Comment:
Maybe it would help reduce confusion to rename `Split` to `ConstBranches`? I was stuck in naming this originally, and it shows in some of the identifiers I picked. 

It's confusing because in`split_telescope` and `split_const_branches` I was trying to follow in an attempt to follow Rust's name for ‘uncons’, [`[T]::split_first`](https://doc.rust-lang.org/std/primitive.slice.html#method.split_first). But then this collides with ‘split’ in the sense of a ‘case-split’.

Also wondering if we could make this type of `split_const_branches` follow [`[T]::split_first`](https://doc.rust-lang.org/std/primitive.slice.html#method.split_first) more closely with something like:

```rs
Option<((Const, ArcValue<'arena>), NextConstBranches<'arena>)>
```

using the following type:

```rs
#[derive(Clone, Debug)]
pub enum NextConstBranches<'arena> {
    Branches(ConstBranches<'arena>),
    Default(Closure<'arena>),
}
```

I dunno if that's better though - just an idea!

Reply:
Ahh interesting. That explanation does connect some dots for me. Perhaps `Split` could also be `CaseSplit` but `ConstBranches` is does seem clearer.

> Also wondering if we could make this type of `split_const_branches` follow [`[T]::split_first`](https://doc.rust-lang.org/std/primitive.slice.html#method.split_first) more closely with something like:
> 
> ```rust
> Option<((Const, ArcValue<'arena>), NextConstBranches<'arena>)>
> ```

This also makes more sense now. I didn't understand in the original code why the tuples were nested versus using a 3-tuple. Perhaps it's worth giving `(Const, ArcValue<'arena>)` a name, even if only an alias like `Branch` or `ConstBranch` to clean up the nesting a bit?

Reply:
Yeah good idea about giving it a name!

Reply:
I had a go at refactoring this to use

```rust
Option<((Const, ArcValue<'arena>), NextConstBranches<'arena>)>
```

but it got a bit messy. Perhaps I misunderstood something but I decided to mostly leave it alone. I did change 

```rust
SplitConstBranches::Const(Const, ArcValue<'arena>, Split<'arena>)
```

to

```rust
SplitConstBranches::Branch(Branch<'arena>, Split<'arena>)
```

though, where `Branch<'arena>` is `type Branch<'arena> = (Const, ArcValue<'arena>);`

Reply:
Yeah that's fair enough - I might have a go later if that's ok.

Eventually might be good to rename `Split` to `Branches` but it's probably good to take one thing at a time.

Reply:
> Yeah that's fair enough - I might have a go later if that's ok.

By all means go ahead 👍